go/api/base_client.go

```diff
@@ -0,0 +1,74 @@
+// Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0
+
+package api
+
+// #cgo LDFLAGS: -L../target/release -lglide_rs
+// #include "../lib.h"
+//
+// void successCallback(uintptr_t channelPtr, char *message);
+// void failureCallback(uintptr_t channelPtr, char *errMessage, RequestErrorType errType);
+import "C"
+
+import (
+	"errors"
+	"unsafe"
+
+	"github.com/aws/glide-for-redis/go/glide/protobuf"
+	"google.golang.org/protobuf/proto"
+)
+
+//export successCallback
+func successCallback(channelPtr C.uintptr_t, cResponse *C.char) {
+	// TODO: Implement when we implement the command logic
+}
+
+//export failureCallback
+func failureCallback(channelPtr C.uintptr_t, cErrorMessage *C.char, cErrorType C.RequestErrorType) {
+	// TODO: Implement when we implement the command logic
+}
+
+type connectionRequestConverter interface {
+	toProtobuf() *protobuf.ConnectionRequest
+}
+
+type baseClient struct {
+	coreClient unsafe.Pointer
+}
+
+func createClient(converter connectionRequestConverter) (*baseClient, error) {
+	request := converter.toProtobuf()
+	msg, err := proto.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	byteCount := len(msg)
+	requestBytes := C.CBytes(msg)
+	cResponse := (*C.struct_ConnectionResponse)(
+		C.create_client(
+			(*C.uchar)(requestBytes),
+			C.uintptr_t(byteCount),
+			(C.SuccessCallback)(unsafe.Pointer(C.successCallback)),
+			(C.FailureCallback)(unsafe.Pointer(C.failureCallback)),
+		),
+	)
+	defer C.free_connection_response(cResponse)
+
+	cErr := cResponse.error_message
+	if cErr != nil {
+		return nil, goError(cResponse.error_type, cResponse.error_message)
+	}
+
+	return &baseClient{cResponse.conn_ptr}, nil
+}
+
+// Close terminates the client by closing all associated resources.
+func (client *baseClient) Close() error {
+	if client.coreClient == nil {
+		return errors.New("close called on uninitialized or already closed client")
```

Comment:
Do we want to create a generic GlideError type for errors like this? Or just use the generic error?

Reply:
Also one negative thing about Close returning an error is that the defer statement is now more complex. If the Close return signature didn't include an error type you could just do `defer client.Close()` after opening the connection. With the error return type, you will get a warning that the error return value is ignored. Instead, you have to do
```go
defer func() {
    err := client.Close()
    if err != nil {
        // ...
    }
}()
```
It would be ideal if we didn't return an error here but maybe that's not an option

Reply:
when close is called for the other clients they do not check whether they were previously closed, they just terminate all resources regardless. I think I'll remove the error and just return without doing anything if baseClient is nil.